### ISSUE TITLE
Make this library Dependency Injection Friendly

### ISSUE THREAD

**Issue by dansiegel** [Azure/azure-mobile-apps-net-client#300]
it really is a horrible pattern to have to do `new SomeType( "aString" )`. Things like the Data Store or connection should be configured through Interfaces that provide any properties that may be required like the path to the database or the Mobile Backend.

**Comment by adrianhall** [comment]
Hey Dan,

I'm not sure I'm following you.   We already do everything through interfaces.  The IMobileServiceClient is pretty much the only one that you need, and the constructor always takes a URI.  What is your proposal for the pattern and how would this be used?  I commonly create a wrapper service manager to do this for me, and then DI the service manager.

How would you actually see this working that's different?

**Comment by dansiegel** [comment]
@adrianhall the constructor for the `MobileServiceClient` currently takes a [string](https://github.com/Azure/azure-mobile-apps-net-client/blob/master/src/Microsoft.Azure.Mobile.Client/MobileServiceClient.cs#L199) or [Uri](https://github.com/Azure/azure-mobile-apps-net-client/blob/master/src/Microsoft.Azure.Mobile.Client/MobileServiceClient.cs#L215) for the mobile service API along with the "optional" `HttpMessageHandler` array. The result is that you cannot directly register the MobileServiceClient with DI unless you were to create a new instance and register it as a singleton. What I would propose would be something like the following:

```cs
public interface IMobileServiceClientOptions
{
    Uri MobileAppUri { get; }
    IEnumerable<HttpMessageHandler> MessageHandlers { get; }
}

public class MobileServiceClient : IMobileServiceClient
{
     public MobileServiceClient(IMobileServiceClientOptions options)
     { 
     }
}
```

Since the MobileServiceClient is already making a major version bump I would suggest just removing the existing constructors.

The would then open things up for a better DI pattern like:

```cs
public class Constants
{
     public const string MobileAppUri = "https://foo.bar.com";
}

public class MyAppMobileServiceClientOptions : IMobileServiceClientOptions
{
    public Uri MobileAppUri { get { return new Uri(Constants.MobileAppUri, UriKind.Absolute); } }
    public IEnumerable<HttpMessageHandler> MessageHandlers { get { return new HttpMessageHandler[]; } }
}

public class App : Application
{
    private void RegisterServices()
    {
        Container.RegisterType<IMobileServiceClient, MobileServiceClient>();
        Container.RegisterType<IMobileServiceClientOptions, MyAppMobileServiceClientOptions>();
    }
}

public class AzureAppService
{
    public AzureAppService(IMobileServiceClient client)
    {
    }
}
```

**Comment by adrianhall** [comment]
You should be creating the MobileServiceClient as a singleton anyway - at least, that is the best practice and recommendation.  If you don't, you lose stateful properties like authentication and push registration.

We tend to advise people to do something like this:

```
public interface ICloudService
{
    public ICloudTable<T> GetTable<T>();
}

public class AzureCloudService : ICloudService
{
    const string uri = "https://foo.azurewebsites.net";
    private MobileServiceClient client;

    public AzureCloudService() 
    {
        client = new MobileServiceClient(uri);
    }

    public ICloudTable<T> GetTable() 
    {
        return new AzureCloudTable<T>(client);
    }
}
```

You can then DI or do a simple ServiceLocator for the AzureCloudService / ICloudService.

**Comment by dansiegel** [comment]
@adrianhall totally understand that, and that is the pattern I have followed in the past as well. I realize that my sample above wasn't being clear about marking the IMobileServiceClient as a Singleton. That said, IMHO, I think it's a better practice to have the MobileServiceClient constructed as a Singleton by the Container rather than doing `new MobileServiceClient(uri)`, and especially with that in the ctor of the AzureCloudService

**Comment by adrianhall** [comment]
I think I'm being incredibly dense or something then.  I honestly don't see the difference or the benefit that your suggestion brings to the code.  If I can wrap the concrete mobileserviceclient into another class that is dependency injectable, what benefit does making mobileserviceclient dependency injectable bring?

**Comment by dansiegel** [comment]
@adrianhall, I recognize there is always more than one way to accomplish something. For example you could do:

```cs
Container.RegisterInstance<IMobileServiceClient>(new MobileServiceClient(uri))
```

Depending on the container you use you may also use some sort of parameter injection to circumvent the issue. That said, relying on interfaces over concrete and primitive types is always a better practice, particularly for testability. Just as an example (and again IMHO) using the `AzureCloudService` you gave the example of above, that would actually be a bad practice because you are now making the service explicitly reliant on the MobileServiceClient rather than IMobileServiceClient which ends up reducing it's testability. 

```cs
public class AzureCloudService : ICloudService
{
    private IMobileServiceClient _client { get; }

    public AzureCloudService(IMobileServiceClient client) 
    {
        _client = client;
    }

    public ICloudTable<T> GetTable() 
    {
        return new AzureCloudTable<T>(_client);
    }
}
```

Don't get me wrong, if I were going to pick a fight on something for the MobileServiceClient it would be to get the Server packages updated to support AspNetCore. And then to see the Xamarin Connected App templates provide full Online/Offline sync with Token Authentication out of the box that's fully buildable on Mac and PC. I just think, updating the design pattern here, would provide for better design practices.

**Comment by adrianhall** [comment]
We merged your PR into the master branch.  It will be released as v4.2.0 within the next couple of weeks.

**Comment by dansiegel** [comment]
thanks @adrianhall here I thought that this project had been abandoned.
